### PR TITLE
firefox: Enable crash reporter by default.

### LINKED ISSUE #107889: Firefox: about:crashes don't work
**Describe the bug**
Firefox crashed because my disk was full. I wanted to submit a crash report, but the site the documentation mentions don't exist.

https://support.mozilla.org/en-US/kb/mozillacrashreporter

**To Reproduce**
1. open `about:crashes` in firefox

**Expected behavior**
firefox works like an official release

**Screenshots**
Screenshot from 2020-12-29 17-21-55

**Additional context**

**Notify maintainers**
@eelco

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.1, NixOS, 21.03.git.173696f216f (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"home-manager-20.09, nixos-20.09.2344.a3a3dda3bac, nixos-hardware, nixos-unstable-21.03pre257780.e9158eca70a"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by kevincox**
This makes it easier to report bugs upstream and ensures that NixOS user crashes are fixed.

Based on my testing these reports weren't automatically submitted anyways so the privacy concern is minimal. For my crashes it both asked my if a crash report should be produced after a crash, and required manual sending. Although possibly the report world eventually be sent automatically.

Fixes https://github.com/NixOS/nixpkgs/issues/107889

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kevincox** [comment]
Result of `nixpkgs-review pr 171900` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>icedtea8_web</li>
    <li>icedtea_web</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>adoptopenjdk-icedtea-web</li>
    <li>firefox</li>
    <li>firefox-esr</li>
    <li>firefox-esr-91-unwrapped</li>
    <li>firefox-esr-wayland</li>
    <li>firefox-unwrapped</li>
    <li>firefox-wayland</li>
    <li>slimerjs</li>
    <li>thunderbird</li>
    <li>thunderbird-unwrapped</li>
    <li>thunderbird-wayland</li>
  </ul>
</details>

**Review by lovesegfault — APPROVED** [review]
I favor this.

**Comment by mweinelt** [comment]
I think we previously didn't enable it by default because we had no debug symbols. This changed quite recently and if we know produce meaningful reports then we should ideally enable the crash reporter.

cc @oxalica, who I think has some experience with the crash reporter.

**Comment by kevincox** [comment]
Here is a crash report: https://crash-stats.mozilla.org/report/index/b15f3c00-f7b0-43f3-85f6-799e10220507#tab-details

It doesn't seem to have symbols. This is strange because running `coredumpctl gdb` has symbols in the stack trace. I'm not completely up to date with our debugging magic though.

Even still I think this is probably useful to have enabled. It at least gives the error messages and what modules are failing. This can probably be correlated with other reports and used for initial triage. And hopefully serves as a base if we get the symbols wired up properly.

**Comment by mweinelt** [comment]
I think we should look into why debug symbols aren't working, as they are crucial to understand the code path that led to a crash.

And on second thought, we should probably also tie the crash report functionality to whether debug symbols are enabled.

**Review by davidak — APPROVED** [review]
Change looks good

I support this change

Not built or tested

**Comment by kevincox** [comment]
To me crash reports without debug symbols seem better than nothing at all. You can at lest quantify what crashes are occurring and get some idea of how frequently. Especially if a user attaches their email then maybe Mozilla can reach out for a symbolized back trace. My preference would be to merge this now, then we can hopefully try to fix symbols in a second step (although I don't know anything about this myself). I also don't see a need to tie the default to if debugging symbols are enabled because to me it seems that some reports are still better than no crash reports. But IDK if we have a good contact at mozilla to actually confirm that.

**Comment by S-NA** [comment]
I think the symbols need to be uploaded to the [Mozilla Symbol Server](https://symbols.mozilla.org/) manually. How to do that can be found [here](https://firefox-source-docs.mozilla.org/crash-reporting/uploading_symbol.html).

**Comment by kevincox** [comment]
Hmm, that sounds like it would take a bit if work to integrate into the nixpkgs automatic builds.

**Comment by kevincox** [comment]
@mweinelt what is the verdict here? Do you think we shouldn't merge this until we get symbol uploading working?

Personally I think this is still an improvement and would merge and we can consider doing uploading later. But your call.

**Comment by mweinelt** [comment]
> Without the debug symbols that match your build the crash reports will not contain actionable information.

https://firefox-source-docs.mozilla.org/crash-reporting/uploading_symbol.html

**Comment by kevincox** [comment]
Ok. I'm not going to pursue that at this time. But anyone is welcome to pick it up form here.

**Comment by davidak** [comment]
@mweinelt can you create a new issue to setup the automation to send it on package updates and tag the maintainers?

That seem out of scope for this.

**Comment by collares** [comment]
As stated in https://github.com/NixOS/nixpkgs/pull/146784#issuecomment-974700781, fixing crash reporting requires that someone packages https://github.com/mozilla/dump_syms and uses it to upload debug symbols to Mozilla servers on a per-build basis.

**Comment by felixsinger** [comment]
@davidak It would be great if you not just think but also communicate with others before you do any action. Leave decisions to the package maintainers and wait for their approvements, especially when the package maintainer mentioned issues.

**Comment by mweinelt** [comment]
The `buildMozillaMach` packaging function creates a symbols output since #172710.

I've created https://github.com/mweinelt/mach-symbols-upload/ to faciliate automatic uploads from NixOS channels to https://symbols.mozilla.org. 

A few hours ago the first uploads for firefox, firefox-esr and thunderbird (both aarch64-linux and x86_64-linux) happened.

Would be delighted to see a new crash report to make sure everything works.

**Comment by kevincox** [comment]
I got a crash report from thunderbird here with no symbols visible: https://crash-stats.mozilla.org/report/index/8669b1e7-c84e-45e3-ab28-8a9d60220524.

**Comment by mweinelt** [comment]
Well, that's Thunderbird 78.10.2, which isn't even supported upstream anymore. I'm not even sure what version of NixOS you're running. Debug symbols are only available for builds that appear on `nixos-unstable`, `nixpkgs-unstable` currently. Will be available for 22.05 channels soon, never for 21.11.

**Comment by kevincox** [comment]
Hmm, interesting. I'm using 91.9 now. I wonder how I got a crash report from that version. The about:crashes page only shows the submitted date now but I seem to recall that the crash date was reported as only a few days ago. But I don't know why I would be running 78 a few days ago.

However a newer crash does appear to have symbols, so I guess I just had a very old crash report lying around somehow? 🤷 https://crash-stats.mozilla.org/report/index/e2ffcb01-5157-48cf-895f-b8df50220524

Either way, looks like it is working at least for thunderbird.

**Pull request merged**
